mantidproject/mantid
# Correctly format FullProf output in SaveReflections

**Description of work.**
There were several issues with SaveReflections not formatting FullProf output that have been addressed
- Add a title line on first line
- Set multiplicity of reflection (Cod) to be one by default - fixed bug where went from [1..npeaks]
- When modulation vectors are present then they need to be in the header - two rows per q (i.e. one each for +/-q) preceded by an index number starting from 1. There is then instead of having a column per q (as in Jana) there is just one column and it references the q index.
- If it looks like constant wavelength then put wavelength in heading (default is 0 which is ignored for TOF Laue)

In addition I have fixed the bug which led to incorrect HKL determined for parent peak of satellites: previously the HKL of the parent was determined casting to int but this fails for modulations with components >=0.5 (note that for non-primitive cells the modulation can be >0.5). It was this change that necessitated changing the Jana test files as well as FullPRof (now exported peaks have same HKL in both formats!).


**To test:**

(1) Run this script
```
LoadNexusProcessed(Filename='TOPAZ_3007.peaks.nxs', OutputWorkspace='peaks')
LoadIsawUB(InputWorkspace='peaks', Filename='TOPAZ_3007.mat')
IndexPeaks(PeaksWorkspace='peaks')
SaveReflections('peaks', Filename='test_fullprof.hkl', Format='FullProf') # parent
PredictFractionalPeaks(Peaks='peaks', RequirePeaksOnDetector=False, ModVector1='0.2,0,0', MaxOrder=1, FracPeaks='frac')
SaveReflections('frac', Filename='test_fullprof_mod.hkl', Format='FullProf') # modulated
```
(2) Check that the output looks as it should (see description)
(3) Change the modulation to be 0.5 and re-run
```
PredictFractionalPeaks(Peaks='peaks', RequirePeaksOnDetector=False, ModVector1='0.5,0,0', MaxOrder=1, FracPeaks='frac_com')
SaveReflections('frac_com', Filename='test_fullprof_mod_commensurate.hkl', Format='FullProf') # modulated commensurate
```
The parent Bragg peak HKL should be the same as before




Fixes #31004 and #31005





---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.
